File: string-paradigm-test-generator/src/main/java/org/jesperancinha/string/paradigm/inputgenerator/InputGeneratorOptions.java
package org.jesperancinha.string.paradigm.inputgenerator;

import org.kohsuke.args4j.Option;

public class InputGeneratorOptions {
	private static final String SPTG_ID_TXT = "sptg_id.txt";

	private static final String SPTG_MOVIE_TXT = "sptg_movie.txt";

	private static final String SPTG_YEAR_TXT = "sptg_year.txt";

	private static final String SPTG_LOCATION_TXT = "sptg_location.txt";

	private static final String SPTG_GENRE_TXT = "sptg_genre.txt";

	private static final String SPTG_DIRECTOR_TXT = "sptg_director.txt";

	private static final String SPTG_PRODUCER_TXT = "sptg_producer.txt";

	private static final String SPTG_ACTOR_TXT = "sptg_actor.txt";

	private static final String SPTG_ACTRESS_TXT = "sptg_actress.txt";

	private static final String SPTG_STUNT_MAN_TXT = "sptg_stunt_man.txt";

	private static final String SPTG_STUNT_WOMAN_TXT = "sptg_stunt_woman.txt";

	private static final String SPTG_SONG_TXT = "sptg_song.txt";

	@Option(name = "-p", aliases = "--number-of-parents", required = true)
	private int numberOfParents;

	@Option(name = "-n", aliases = "--number-of-elements-per-node", required = true)
	private int numberOfElements;

	@Option(name = "-d", aliases = "--directory", required = false)
	private String folder = "/tmp";

	@Option(name = "-f", aliases = "--filename", required = true)
	private String filename;

	/**
	 * A plain text file with the film id data
	 */
	@Option(name = "-mid", aliases = "--movie-id", required = false)
	private String movieIdFile = SPTG_ID_TXT;

	/**
	 * A plain text file with the film name data
	 */
	@Option(name = "-m", aliases = "--m", required = false)
	private String movieFile = SPTG_MOVIE_TXT;

	/**
	 * A plain text file with the film year data
	 */
	@Option(name = "-my", aliases = "--movie-year", required = false)
	private String movieYearFile = SPTG_YEAR_TXT;

	/**
	 * A plain text file with the film location data
	 */
	@Option(name = "-ml", aliases = "--movie-location", required = false)
	private String movieLocationFile = SPTG_LOCATION_TXT;

	/**
	 * A plain text file with the film location data
	 */
	@Option(name = "-mg", aliases = "--movie-genre", required = false)
	private String movieGenreFile = SPTG_GENRE_TXT;

	/**
	 * A plain text file with the film director data
	 */
	@Option(name = "-md", aliases = "--movie-director", required = false)
	private String movieDirectorFile = SPTG_DIRECTOR_TXT;

	/**
	 * A plain text file with the film producer data
	 */
	@Option(name = "-mp", aliases = "--movie-producer", required = false)
	private String movieProducerFile = SPTG_PRODUCER_TXT;

	/**
	 * A plain text file with the actor data
	 */
	@Option(name = "-mar", aliases = "--movie-actor", required = false)
	private String movieActorFile = SPTG_ACTOR_TXT;

	/**
	 * A plain text file with the actress data
	 */
	@Option(name = "-mas", aliases = "--movie-actress", required = false)
	private String movieActressFile = SPTG_ACTRESS_TXT;

	/**
	 * A plain text file with the stuntman data
	 */
	@Option(name = "-msman", aliases = "--movie-stunt-man", required = false)
	private String movieStuntManFile = SPTG_STUNT_MAN_TXT;

	/**
	 * A plain text file with the stuntwoman data
	 */
	@Option(name = "-mswoman", aliases = "--movie-stunt-woman", required = false)
	private String movieStuntWomanFile = SPTG_STUNT_WOMAN_TXT;

	/**
	 * A plain text file with the song data
	 */
	@Option(name = "-msong", aliases = "--movie-song", required = false)
	private String movieSongFile = SPTG_SONG_TXT;

	public int getNumberOfElements() {
		return numberOfElements;
	}

	public String getFolder() {
		return folder;
	}

	public String getFilename() {
		return filename;
	}

	public String getMovieFile() {
		return movieFile;
	}

	public String getMovieActorFile() {
		return movieActorFile;
	}

	public String getMovieActressFile() {
		return movieActressFile;
	}

	public String getMovieIdFile() {
		return movieIdFile;
	}

	public String getMovieDirectorFile() {
		return movieDirectorFile;
	}

	public String getMovieGenreFile() {
		return movieGenreFile;
	}

	public String getMovieLocationFile() {
		return movieLocationFile;
	}

	public String getMovieProducerFile() {
		return movieProducerFile;
	}

	public String getMovieSongFile() {
		return movieSongFile;
	}

	public String getMovieStuntManFile() {
		return movieStuntManFile;
	}

	public String getMovieStuntWomanFile() {
		return movieStuntWomanFile;
	}

	public String getMovieYearFile() {
		return movieYearFile;
	}

	public int getNumberOfParents() {
		return numberOfParents;
	}
}

File: string-paradigm-test-generator/src/main/kotlin/org/jesperancinha/string/paradigm/inputgenerator/InputGeneratorOptions.kt
package org.jesperancinha.string.paradigm.inputgenerator

import picocli.CommandLine.Command
import picocli.CommandLine.Option
import java.io.*
import java.nio.charset.Charset
import java.util.*
import java.util.concurrent.Callable
import java.util.logging.Logger


@Command(
    name = "input-generator", mixinStandardHelpOptions = true, version = ["0.0.0"],
    description = ["Generates input test files"]
)
open class InputGeneratorOptions : Callable<Int> {
    @Option(names = ["-p,--number-of-parents"], required = true)
    open val numberOfParents = 0

    @Option(names = ["-n,--number-of-elements-per-node"], required = true)
    open val numberOfElements = 0

    @Option(names = ["-d,--directory"], required = false)
    val folder = "/tmp"

    @Option(names = ["-f,--filename"], required = true)
    open val filename: String? = null

    /**
     * A plain text file with the film id data
     */
    @Option(names = ["-mid,--movie-id"], required = false)
    val movieIdFile = SPTG_ID_TXT

    /**
     * A plain text file with the film name data
     */
    @Option(names = ["-m,--m"], required = false)
    val movieFile = SPTG_MOVIE_TXT

    /**
     * A plain text file with the film year data
     */
    @Option(names = ["-my,--movie-year"], required = false)
    val movieYearFile = SPTG_YEAR_TXT

    /**
     * A plain text file with the film location data
     */
    @Option(names = ["-ml,--movie-location"], required = false)
    val movieLocationFile = SPTG_LOCATION_TXT

    /**
     * A plain text file with the film location data
     */
    @Option(names = ["-mg,--movie-genre"], required = false)
    val movieGenreFile = SPTG_GENRE_TXT

    /**
     * A plain text file with the film director data
     */
    @Option(names = ["-md,--movie-director"], required = false)
    val movieDirectorFile = SPTG_DIRECTOR_TXT

    /**
     * A plain text file with the film producer data
     */
    @Option(names = ["-mp,--movie-producer"], required = false)
    val movieProducerFile = SPTG_PRODUCER_TXT

    /**
     * A plain text file with the actor data
     */
    @Option(names = ["-mar,--movie-actor"], required = false)
    val movieActorFile = SPTG_ACTOR_TXT

    /**
     * A plain text file with the actress data
     */
    @Option(names = ["-mas,--movie-actress"], required = false)
    val movieActressFile = SPTG_ACTRESS_TXT

    /**
     * A plain text file with the stuntman data
     */
    @Option(names = ["-msman,--movie-stunt-man"], required = false)
    val movieStuntManFile = SPTG_STUNT_MAN_TXT

    /**
     * A plain text file with the stuntwoman data
     */
    @Option(names = ["-mswoman,--movie-stunt-woman"], required = false)
    val movieStuntWomanFile = SPTG_STUNT_WOMAN_TXT

    /**
     * A plain text file with the song data
     */
    @Option(names = ["-msong,--movie-song"], required = false)
    val movieSongFile = SPTG_SONG_TXT

    override fun call(): Int {
        val directory = File(folder)
        directory.mkdirs()
        val outfile = File(directory, filename)
        val streamId = getStreamFromFileName(movieIdFile)
        val streamYear = getStreamFromFileName(movieYearFile)
        val streamMovie = getStreamFromFileName(movieFile)
        val streamGenre = getStreamFromFileName(movieGenreFile)
        val streamLocation = getStreamFromFileName(movieLocationFile)
        val streamDirector = getStreamFromFileName(movieDirectorFile)
        val streamProducer = getStreamFromFileName(movieProducerFile)
        val streamActress = getStreamFromFileName(movieActressFile)
        val streamStuntWoman = getStreamFromFileName(movieStuntWomanFile)
        val streamActor = getStreamFromFileName(movieActorFile)
        val streamStuntMan = getStreamFromFileName(movieStuntManFile)
        val streamSong = getStreamFromFileName(movieSongFile)
        val resultId = getFullFileList(streamId)
        val resultActor = getFullFileList(streamActor)
        val resultActress = getFullFileList(streamActress)
        val resultDirector = getFullFileList(streamDirector)
        val resultGenre = getFullFileList(streamGenre)
        val resultLocation = getFullFileList(streamLocation)
        val resultMovie = getFullFileList(streamMovie)
        val resultProducer = getFullFileList(streamProducer)
        val resultSong = getFullFileList(streamSong)
        val resultStuntMan = getFullFileList(streamStuntMan)
        val resultStuntWoman = getFullFileList(streamStuntWoman)
        val resultYear = getFullFileList(streamYear)
        val maxNumberOfElements = numberOfElements
        val nIds = numberOfParents
        BufferedWriter(FileWriter(outfile)).use { writer ->
            for (i in 0 until nIds) {
                var bf = StringBuffer()
                bf.append(getRandomStringFromArray(resultId) + "" + i)
                val nYears = RANDOM.nextInt(maxNumberOfElements) + 1
                val stringYear = bf.toString()
                for (j in 0 until nYears) {
                    bf.append(getRandomStringFromArrayPreColon(resultYear))
                    val nMovies = RANDOM.nextInt(maxNumberOfElements) + 1
                    val stringMovie = bf.toString()
                    for (k in 0 until nMovies) {
                        bf.append(getRandomStringFromArrayPreColon(resultMovie))
                        val nGenres = RANDOM.nextInt(maxNumberOfElements) + 1
                        val stringGenre = bf.toString()
                        for (j2 in 0 until nGenres) {
                            bf.append(getRandomStringFromArrayPreColon(resultGenre))
                            val nLocation = RANDOM.nextInt(maxNumberOfElements) + 1
                            val stringLocation = bf.toString()
                            for (k3 in 0 until nLocation) {
                                bf.append(getRandomStringFromArrayPreColon(resultLocation))
                                val nDirector = RANDOM.nextInt(maxNumberOfElements) + 1
                                val stringDirector = bf.toString()
                                for (k2 in 0 until nDirector) {
                                    bf.append(getRandomStringFromArrayPreColon(resultDirector))
                                    val nProducer = RANDOM.nextInt(maxNumberOfElements) + 1
                                    val stringProducer = bf.toString()
                                    for (l in 0 until nProducer) {
                                        bf.append(getRandomStringFromArrayPreColon(resultProducer))
                                        val nActress = RANDOM.nextInt(maxNumberOfElements) + 1
                                        val stringActress = bf.toString()
                                        for (l2 in 0 until nActress) {
                                            bf.append(getRandomStringFromArrayPreColon(resultActress))
                                            val nStuntWoman = RANDOM.nextInt(maxNumberOfElements) + 1
                                            val stringStuntWoman = bf.toString()
                                            for (m in 0 until nStuntWoman) {
                                                bf.append(getRandomStringFromArrayPreColon(resultStuntWoman))
                                                val nActor = RANDOM.nextInt(maxNumberOfElements) + 1
                                                val stringActor = bf.toString()
                                                for (m2 in 0 until nActor) {
                                                    bf.append(getRandomStringFromArrayPreColon(resultActor))
                                                    val nStuntMan = RANDOM.nextInt(maxNumberOfElements) + 1
                                                    val stringStuntMan = bf.toString()
                                                    for (n in 0 until nStuntMan) {
                                                        bf.append(getRandomStringFromArrayPreColon(resultStuntMan))
                                                        val nSong = RANDOM.nextInt(maxNumberOfElements) + 1
                                                        val stringSong = bf.toString()
                                                        for (n2 in 0 until nSong) {
                                                            bf.append(getRandomStringFromArrayPreColon(resultSong))
                                                            writer.write(bf.toString())
                                                            writer.write("\n")
                                                            bf = StringBuffer()
                                                            bf.append(stringSong)
                                                        }
                                                        bf = StringBuffer()
                                                        bf.append(stringStuntMan)
                                                    }
                                                    bf = StringBuffer()
                                                    bf.append(stringActor)
                                                }
                                                bf = StringBuffer()
                                                bf.append(stringStuntWoman)
                                            }
                                            bf = StringBuffer()
                                            bf.append(stringActress)
                                        }
                                        bf = StringBuffer()
                                        bf.append(stringProducer)
                                    }
                                    bf = StringBuffer()
                                    bf.append(stringDirector)
                                }
                                bf = StringBuffer()
                                bf.append(stringLocation)
                            }
                            bf = StringBuffer()
                            bf.append(stringGenre)
                        }
                        bf = StringBuffer()
                        bf.append(stringMovie)
                    }
                    bf = StringBuffer()
                    bf.append(stringYear)
                }
            }
        }
        LOG.info("Completed test file generation")
         return 0
    }

    fun getStreamFromFileName(filename: String?): InputStream {
        return javaClass.getResourceAsStream(filename)
    }

    @Throws(IOException::class)
    fun getFullFileList(io: InputStream?): List<String> {
        val list: MutableList<String> = ArrayList()
        BufferedReader(InputStreamReader(io, Charset.defaultCharset())).use { br ->
            var testRead: String? = br.readLine()
            while (testRead != null) {
                list.add(testRead)
                testRead = br.readLine()
            }
        }
        return list
    }


    companion object {
        private const val SPTG_ID_TXT = "sptg_id.txt"
        private const val SPTG_MOVIE_TXT = "sptg_movie.txt"
        private const val SPTG_YEAR_TXT = "sptg_year.txt"
        private const val SPTG_LOCATION_TXT = "sptg_location.txt"
        private const val SPTG_GENRE_TXT = "sptg_genre.txt"
        private const val SPTG_DIRECTOR_TXT = "sptg_director.txt"
        private const val SPTG_PRODUCER_TXT = "sptg_producer.txt"
        private const val SPTG_ACTOR_TXT = "sptg_actor.txt"
        private const val SPTG_ACTRESS_TXT = "sptg_actress.txt"
        private const val SPTG_STUNT_MAN_TXT = "sptg_stunt_man.txt"
        private const val SPTG_STUNT_WOMAN_TXT = "sptg_stunt_woman.txt"
        private const val SPTG_SONG_TXT = "sptg_song.txt"

        private val LOG = Logger.getLogger(InputGeneratorOptions::class.java.name)
        val RANDOM = Random()

        private fun getRandomStringFromArrayPreColon(results: List<String>): String {
            return ";" + getRandomStringFromArray(results)
        }

        private fun getRandomStringFromArray(results: List<String>): String {
            return results[RANDOM.nextInt(results.size)]
        }
    }


}